pymmcore_plus/core/_mmcore_plus.py

```diff
@@ -283,9 +285,18 @@ def getCameraChannelNames(self) -> Tuple[str, ...]:
             for i in range(self.getNumberOfCameraChannels())
         )
 
-    def run_mda(self, sequence: MDASequence) -> None:
+    def run_mda(self, sequence: MDASequence, writer: MDAWriter = None) -> None:

```

Comment:
I'm not in love with having run_mda take a new parameter, i can imagine a writer being just one of many things that may want to peek in on the running experiment and do something with it.  This also makes it hard to add, two writers (who knows why you'd want to do it, but it's not a ridiculous concept...).

I still think I'd prefer the concept of a [temporary signal listener](https://github.com/tlambert03/pymmcore-plus/pull/29#discussion_r679283108)
any thoughts there?

I also think we should just improve the `axis_order` getting on the `MDASequence` object as needed, rather than adding convenience methods here (`get_axis_order`) and then passing `sequence.shape`, `get_axis_order(sequence)`, _and_ `sequence` to the writer... when it seems like we could just pass `sequence`.

Reply:
> I'm not in love with having run_mda take a new parameter, 

That's sensible but I don't quite see how that would work with your earlier suggestion about narrowed scope (unless the the temporary listener did some magic to find the correct methods).:

> rather than a general event follower: do you think we could narrow down the scope of the writer protocol? It seems like it probably needs three methods:

    initialize: provide knowledge of the axis names involved, and the expected shape of the final data
    add_frame: provide a new 2D image, and slice object corresponding to it's position in the shape of the full data
    finalize: let it know it's done.


> I also think we should just improve the axis_order getting on the MDASequence object as needed, rather than adding convenience methods here (get_axis_order) and then passing sequence.shape, get_axis_order(sequence), and sequence to the writer... when it seems like we could just pass sequence.

Agreed, I'll try to open a PR on useq